ratbag_emu/device.py

```diff
@@ -0,0 +1,272 @@
+            endpoint.send(endpoint.create_report(action, 0x11))
+
+    def simulate_action(self, action: Dict[str, Any], type: int = None):
+        '''
+        Simulates action
+
+        Translates physical values according to the device properties and
+        converts action into HID reports.
+
+        :param action:  high-level action
+        :param type:    HID report type
+        '''
+
+        packets: Dict[int, EventData] = {}
+
+        report_count = int(round(ms2s(action['duration']) * self.report_rate))
+
+        if not report_count:
+            report_count = 1
+
+        # XY movement
+        if action['type'] == 'xy':
+            # We assume a linear motion
+            pixel_buffer = {}
+            step = {}
+
+            '''
+            Initialize pixel_buffer, real_pixel_buffer and step for X and Y
+
+            pixel_buffer holds the ammount of pixels left to send (kinda,
+            read below).
+
+            We actually have two variables for this, real_pixel_buffer and
+            pixel_buffer. pixel_buffer mimics the user movement and
+            real_pixel_buffer holds true number of pixels left to send.
+            When using high report rates (ex. 1000Hz) we usually don't
+            have a full pixel to send, that's why we need two variables. We
+            subtract the step to pixel_buffer at each iteration, when the
+            difference between pixel_buffer and real_pixel_buffer is equal
+            or higher than 1 pixel we then send a HID report to the device
+            with that difference (int! we send the int part of the
+            difference) and update real_pixel_buffer to include this
+            changes.
+            '''
+            pixel_buffer = self.transform_action(action['data'])
+
+            for attr in ['x', 'y']:
+                assert pixel_buffer[attr] <= 127 * report_count
+                step[attr] = pixel_buffer[attr] / report_count
+
+            real_pixel_buffer = copy.deepcopy(pixel_buffer)
+
+            i = 0
+            while real_pixel_buffer['x'] != 0:
+                if i not in packets:
+                    packets[i] = EventData()
+
+                for attr in ['x', 'y']:
+                    pixel_buffer[attr] -= step[attr]
+                    diff = int(round(real_pixel_buffer[attr] -
+                               pixel_buffer[attr]))
+                    # FIXME: Read max size from the report descriptor
+                    '''
+                    The max is 127, if this happens we need to leave the
+                    excess in the buffer for it to be sent in the next
+                    report
+                    '''
+                    if abs(diff) >= 1:
+                        if abs(diff) > 127:
+                            diff = 127 if diff > 0 else -127
+                        setattr(packets[i], attr, diff)
+                        real_pixel_buffer[attr] -= diff
+                i += 1
+
```

Comment:
This code segment is too big to be embedded in an other function. Please extract it in an other function `simulate_action_xy`.

Same for the one below: `simulate_action_button`